File: maze.py
import random

N, M = 15, 25
a, b = -1, -1
grid = [['.' for x in range(M)] for y in range(N)]

n = random.randint(0, N - 1)
n2 = random.randint(0, N - 1)


# This function prints the grid of Gomoku as the game progresses
def print_grid():
    print('__' + '___' * M + '__')
    for i in range(N):
        if i == n:
            print(end="   ")
        else:
            print(end='|  ')
        for j in range(M):
            print(grid[i][j], end='  ')
        if i == n2:
            print(end=" ")
        else:
            print(end='|')
        print()
    print('|_' + '___' * M + '_|')


# This function checks if the game has a win state or not
def check_win():
    if a == n2 and b + 1 == M:
        return True
    else:
        return False


# This function checks if given position is valid or not
def check_valid_direction(i):
    if i == 1 or i == 2 or i == 3 or i == 4:
        return True
    else:
        return False


# This function generates man
def generate_man():
    global a
    global b
    a = n
    b = 0
    grid[a][b] = 'M'


def check_available_move(i):
    if i == 3:
        if b + 1 > M - 1:
            return False  # right
    elif i == 1:
        if a - 1 < 0:
            return False  # up
    elif i == 2:
        if a + 1 > N - 1:
            return False  # down
    elif i == 4:
        if b - 1 < 0:
            return False  # left
    return True


# This function checks if the game has a tie state or not for the given mark
def check_lose(i):
    if i == 3:
        if grid[a][b + 1] == 'G':
            return True
    elif i == 1:
        if grid[a - 1][b] == 'G':
            return True
    elif i == 2:
        if grid[a + 1][b] == 'G':
            return True
    elif i == 4:
        if grid[a][b - 1] == 'G':
            return True
    return False


def check_wall(i):
    if i == 3:
        if grid[a][b + 1] == '|' or grid[a][b + 1] == '-':
            return True
    elif i == 1:
        if grid[a - 1][b] == '|' or grid[a][b + 1] == '-':
            return True
    elif i == 2:
        if grid[a + 1][b] == '|' or grid[a][b + 1] == '-':
            return True
    elif i == 4:
        if grid[a][b - 1] == '|' or grid[a][b + 1] == '-':
            return True
    return False


def move(i):
    global a
    global b

    grid[a][b] = '.'
    if i == 3:  # right
        grid[a][b + 1] = 'M'
        b += 1

    elif i == 1:  # up

        grid[a - 1][b] = 'M'
        a -= 1

    elif i == 2:  # down

        grid[a + 1][b] = 'M'
        a += 1

    elif i == 4:  # left

        grid[a][b - 1] = 'M'
        b -= 1


# This function generates ghosts
def generate_():
    f = N * M * 1 // 6
    while f > 0:
        a = random.randint(0, N - 1)
        b = random.randint(0, M - 1)
        while grid[a][b] == 'M' or (a == n2 and b == M - 1) or (a == n2 - 1 and b == M - 1) or (
                a == n2 + 1 and b == M - 1) or (a == n and b == 1) or (a == n - 1 and b == 0) or (
                a == n + 1 and b == 0) or (a == n - 1 and b == 1) or (a == n + 1 and b == 1):
            a = random.randint(0, N - 1)
            b = random.randint(0, M - 1)
        grid[a][b] = '_'
        f -= 1

    # This function generates a wall


def generate_wall():
    f = N * M * 1 // 8
    while f > 0:
        a = random.randint(0, N - 1)
        b = random.randint(0, M - 1)
        while grid[a][b] == 'M' or grid[a][b] == '_' or (a == n2 and b == M - 1) or (a == n2 - 1 and b == M - 1) or (
                a == n2 + 1 and b == M - 1) or (a == n and b == 1) or (a == n - 1 and b == 0) or (
                a == n + 1 and b == 0) or (a == n - 1 and b == 1) or (a == n + 1 and b == 1):
            a = random.randint(0, N - 1)
            b = random.randint(0, M - 1)
        grid[a][b] = '|'
        f -= 1


# This function clears the game structures
def grid_clear():
    global grid
    grid = [['.' for x in range(M)] for y in range(N)]

    global a, b
    a, b = -1, -1


# This function reads a valid position input
def read_input():
    i = int(input('Enter Direction : '))
    while not check_valid_direction(i):
        i = int(input('Enter a valid direction: '))
    return i


# MAIN FUNCTION
def play_game():
    print("Maze  Game!")
    print("Welcome...")
    print("============================")
    while True:
        # Prints the grid
        print_grid()
        i = read_input()
        if check_wall(i):
            print("WHATCH OUT you hit a wall !! please ", end=' ')
            i = read_input()
        # Move with the input direction
        move(i)
        # Check if the state of the grid has a win state
        if check_win():
            # Prints the grid
            print_grid()
            # Announcement of the final statement
            print("Congrats, you won!\n")
            break


while True:
    grid_clear()
    generate_man()
    generate_()
    generate_wall()
    play_game()
    c = input('Play Again [Y/N] ')
    if c not in 'yY':
        break

File: maze.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <cmath>
using namespace std;
const int N = 15;  //TWL
const int M = 25; //3RD
int counter=0;
char grid[N][M];

//srand( (unsigned)time( NULL ) );
int n2=rand()%N;
int n=rand()%N;
//This function prints the grid of Gomoku as the game progresses
void print_grid() {

    cout << "\n";
	cout << "__";
    for (int i = 0; i < M; cout << "___", i++);
    cout << "__\n";
    for (int i = 0; i < N; i++) {
            if( i==n ){cout<<"   ";}
        else{cout << "|  ";   }   //left
        for (int j = 0; j < M; j++)
            cout << grid[i][j] << "  ";
            if( i==n2 ){cout<<" \n";}
        else {cout << "|\n";}
        cout << "";
        for (int i = 0; i < M; cout << "", i++);
    }
    cout<<"|"; for (int i = 0; i < M; cout << "___", i++);
    cout << "__|\n";
}

//This function checks if the game has a win state or not
bool check_win(int a, int b ) {
    if(a==n2 && b+1==M){return true;}
    else{return false;}

}


//This function checks if the game has a tie state or not for the given mark
bool check_wall(int i, int a , int b) {
if (i==3){  if(grid[a][b+1]=='|' || grid[a][b+1]=='_' ){return true;}}

else if (i==1){ if(grid[a-1][b]=='|' || grid[a][b+1]=='_' ){return true;}}

else if (i==2){  if(grid[a+1][b]=='|' || grid[a][b+1]=='_' ){return true;}}

else if (i==4){  if(grid[a][b-1]=='|' || grid[a][b+1]=='_' ){return true;}}
  return false;
}



//This function checks if given position is valid or not
bool check_valid_direction(int i) {
    if (i==1 || i==2 || i==3 || i==4 ){
        return true;
        }
    else{return false;}
}

//This function generates pac man
void generate_man(int &a , int &b) {

    a = n;
	b = 0;
    grid[a][b] = 'M';
}

//This function generates a wall
void generate_wall() {
    srand( (unsigned)time( NULL ) );
    int f=N*M*1/6;
    while ( f>0 )
    {
    int a = rand() % N;
	int b = rand() % M;
    while ( grid[a][b] == 'M' || (a==n2 && b==M-1) || (a==n2-1 && b==M-1) || (a==n2+1 && b==M-1) || (a==n && b==1)
                             || (a==n-1 && b==0) || (a==n+1 && b==0)||(a==n-1 && b==1) || (a==n+1 && b==1)) {

        a = rand() % N;
        b = rand() % M;
	}
    grid[a][b] = '|';
    f--;
    }
}

void generate_() {
    srand( (unsigned)time( NULL ) );
    int f=N*M*1/8;
    while ( f>0 )
    {
    int a = rand() % N;
	int b = rand() % M;
    while ( grid[a][b] == 'M' || grid[a][b] == '|' || (a==n-1 && b==0) ||
            (a==n+1 && b==0)||(a==n-1 && b==1) || (a==n+1 && b==1) || (a==n && b==1)
           ||(a==n2 && b==M-1) || (a==n2-1 && b==M-1) || (a==n2+1 && b==M-1))  {
        a = rand() % N;
        b = rand() % M;
	}
    grid[a][b] = '_';
    f--;
    }
}



//Move with the input direction
void move(int i,int &a,int &b) {
    grid[a][b]='.';
    if (i==3){  if(grid[a][b+1]=='+'){counter+=1; grid[a][b+1]='M';   }
                else if(grid[a][b+1]=='.'){ grid[a][b+1]='M';  } b+=1;  } //right

 else if (i==1){ if(grid[a-1][b]=='+'){counter+=1; grid[a-1][b]='M';}
                 else if(grid[a-1][b]=='.'){ grid[a-1][b]='M'; }  a-=1;  }   // up

 else if (i==2){  if(grid[a+1][b]=='+'){counter+=1; grid[a+1][b]='M';}
                    else if(grid[a+1][b]=='.'){ grid[a+1][b]='M'; } a+=1;}  //down

 else if (i==4){  if(grid[a][b-1]=='+'){counter+=1; grid[a][b-1]='M';}
                  else if(grid[a][b-1]=='.'){ grid[a][b-1]='M'; }b-=1;}  }   //left

//This function clears the game structures
void grid_clear() {
     for (int i =0 ; i<N; i++){
            for (int j =0 ; j<M; j++){
                grid[i][j]='.';
            }
        }

}
//This function reads a valid direction
void read_input(int &i,int &a, int &b) {
    cout << "Enter the direction: ";
	cin >> i;
    while (!check_valid_direction(i)) {
        cout << "Enter a valid direction: ";
		cin >> i;
	}
}
//MAIN FUNCTION
void play_game(int &a,int &b) {
    cout << "Pac-Man Game!\n";
    cout << "Welcome...\n";
    cout << "============================\n";
    while (true) {

        //Prints the grid
        print_grid();
        //Read an input from the player
		int i;
		read_input(i,a,b);

        //Check if the state of the grid has a tie state
        /*if (check_lose(i,a,b)) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "GAME OVER!\n";
            break;
        }*/
        if(check_wall(i,a,b)){
            cout<<"WHATCH OUT you hit a wall !! please " ;
            read_input(i,a,b);
        }
        //Move with the input direction
        move(i,a,b);
        //Check if the state of the grid has a win state
        if (check_win(a,b)) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Congrats, you won!\n";
            break;
        }

    }
}
int main() {
    while (true) {
        int a,b;
        grid_clear();
        //
        generate_man(a,b);

        generate_wall();
        generate_();
    	play_game(a,b);
    	char c;
    	cout << "Play Again [Y/N] ";
    	cin >> c;
    	if (c != 'y' && c != 'Y')
    		break;
    }
}
